modules/Preprocessing.py

```diff
@@ -0,0 +1,115 @@
+from nltk.corpus import stopwords
+from nltk.tag import pos_tag
+from nltk.tokenize import word_tokenize
+
+
+# lazy load stopwords
+_stopwords = stopwords.words('english')
+_stopwords.extend(['twitter', 'nt'])
+
+
+def _handle_date(date_string: str):
+    """
+    We receive dates from tweets in the format 'yyyy-m-d'.
+    This function returns a datetime object with proper formatting (yyyy-mm-dd)
+    """
+    # return date(*[int(date) for date in date_string.split(',')]) # sorry we didnt get to use you ;(
+    return datetime.strptime(date_string, '%Y,%m,%d').date()
+
+
+def remove_noise(tweet: str):
+    """
+    Removes noise from the tweets by:
+    Tokenizing (Splits sentences into array of words)
+    Removes hyperlinks with regex
+    Removes special characters (primarily used for emojis) as well as numbers.
+    _____________
+    Is used for cleaning both scraped data as well as cleaning the data for training the model
+    """
+    cleaned_tokens = []
+    tweet_tokens = word_tokenize(tweet)
+    for token, tag in pos_tag(tweet_tokens):
+
+        # remove hyperlinks
+        token = re.sub('(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)', '', token)
+        # remove special chars, numbers, inc emojies
+        token = re.sub("[^A-Za-z]", "", token)
+
+        if tag.startswith("NN"): #noun
+            pos = 'n'
+        elif tag.startswith('VB'): #verb
+            pos = 'v'
+        else: #adjective
+            pos = 'a'
+
+        lemmatizer = WordNetLemmatizer()
+        #lemmatize sentence (bring word from full form to base form) (running -> run)
+        token = lemmatizer.lemmatize(token, pos)
+        # remove empty tokens, punctuations and stopwords
+        token = token.lower().strip()
+        if len(token) > 1 and token not in string.punctuation and token not in _stopwords:
+            cleaned_tokens.append(token)
+    return cleaned_tokens
+
+
+def get_tweet_data(tweets: List[Dict[str, str]]):
+    """
+    This function takes a list of tweets, containing web scraped dicts (in particular raw_text) and grabs useful information from it.
+
+    As of now it looks for hashtags (#), mentions (@) and emojis.
+
+    Following this, it cleans up the data using remove_noise().
+
+    ## Returns
+    Returns the same object with fields:  
+
+    hashtags, mentions, tweet.
+
+    As well as stats for hashtags and mentions
+    >>> tweets, hashtag_stats, mention_stats
+    """
+    # prepare stats_format for
+    hashtag_stats = {}
+    mention_stats = {}
+
+    for tweet in tweets:
+        # prepare format
+        tweet['hashtags'] = tweet.get('hashtags', [])
+        tweet['mentions'] = tweet.get('mentions', [])
+        tweet_text = tweet.get('raw_text')
+        # remove newline characters (necessary to add spaces between words)
+        tweet_text = tweet_text.replace('\n', ' ')
+
+        # check text for hashtags or mentions
+        if (tweet_text != None and '#' or '@' in tweet_text):
```

Comment:
Does this work as intended? Wouldn't you have to write `if (tweet_text != None and '#' in tweet_text or '@' in tweet_text):` (with the extra `in tweet_text` after `#`) for it to work properly or am I stuck with Java's syntax?

Reply:
> Does this work as intended? Wouldn't you have to write
> `if (tweet_text != None and ('#' in tweet_text or '@' in tweet_text)):` (with the extra `in tweet_text` after `#`) for it to work properly or am I stuck with Java's syntax?

If you are then I'm stuck in Java syntax too - i think you are right, but I'm not entirely sure.

Reply:
You are correct. I have updated it to use `any` which should be faster (?) as well as the pythonic way to check multiple values according to some dude on stackoverflow.

https://github.com/Hold-Krykke/PythonExam/pull/8/commits/2b51e95129e71076b5116afd06f6033d2cb1a1d2